enhancements/agent-installer/automated-workflow-kubeconfig.md

```diff
@@ -0,0 +1,253 @@
+Agent ISO-->>Cluster creator: agent.iso
+Agent ISO-->>Assisted Service: certificates
+Assisted Service-->>OpenShift Installer: certificates
+OpenShift Installer-->>Assisted Service: kubeconfig
+OpenShift Installer-->>Bootstrap: ignition
+```
+
+### API Extensions
+
+N/A
+
+### Implementation Details/Notes/Constraints
+
+Note that the modifications to the installer to read certificates from disk
+need to occur in the installer present in the payload of the release being
+installed, i.e. on the `master` branch, not the `agent-installer` branch where
+the agent-based installer is being developed.
+
+The work directory for the installer is a tmpdir created by assisted-service on
+the fly with a randomised name. Therefore it is not possible to pre-populate
+the certificates into the directory, and assisted-service must be involved in
+copying the files.
+
+### Risks and Mitigations
+
+End users will be able to supply their own CA certificates to the installer
+much more easily than they could before. If done incorrectly, this may leave
+the cluster less secure than if the keys and certificates are generated inside
+openshift-installer. To discourage users from relying on this, the ability to
+read the certificates from disk will only be enabled when the environment
+variable `OPENSHIFT_INSTALL_LOAD_CLUSTER_CERTS` is set to the value `true`. The
+assisted-service will set this in the environment when running the installer,
+but otherwise it should not be routinely set.
+
+A copy of the admin kubeconfig will remain accessible via the assisted-service
+API, enabling an attacker with access to the network environment of the cluster
+during bootstrapping to gain access to it. We could safely disable this API
+when the `ASSISTED_SERVICE_CLUSTER_TLS_CERTS_DIR` environment variable is set.
+
+### Drawbacks
+
+The same mechanism will likely not be usable in the interactive flow, because
+we won't know the cluster name (and therefore its API address) at the time of
+generating the agent ISO.
+
+## Design Details
+
+### Open Questions
+
+Is there a better way to configure assisted-service to copy the key/certificate
+files?
+
+### Test Plan
+
+Since the admin kubeconfig is required to verify that the cluster installation
+is completed, this feature is expected to be tested on every CI run.
+
+### Graduation Criteria
+
+N/A
+
+#### Dev Preview -> Tech Preview
+
+N/A
+
+#### Tech Preview -> GA
+
+N/A
+
+#### Removing a deprecated feature
+
+N/A
+
+### Upgrade / Downgrade Strategy
+
+N/A
+
+### Version Skew Strategy
+
+N/A
+
+### Operational Aspects of API Extensions
+
+N/A
+
+#### Failure Modes
+
+N/A
+
+#### Support Procedures
+
+N/A
+
+## Implementation History
+
+The changes to the installer to read the certificates/keys from disk were
+implemented in
+[openshift/installer#5872](https://github.com/openshift/installer/pull/5872)
+and are available in OpenShift 4.11 nightlies from Feature Freeze on.
+
+## Alternatives
```

Comment:
Just one more idea to toss in here:

We could let the CA cert, kubeconfig, and kubeadmin password all still be generated the usual way. Then in the new cluster, run a temporary web service just for making those available. It could require a pre-defined token that is generated by the installer, which might be easier to add to the cluster. The service itself might be run based on including an extra manifest.

There are various strategies for ensuring the service stops after the hand-off is complete, so the service doesn't run forever. For example:

- user accesses a known URL served by the new cluster and presents their token. This could be via a UI or just an API response.
- The access service responds with a limited kubeconfig that can only access a specific namespace and with limited capability.
- The user creates an empty secret with a known name in that namespace.
- The access service sees the secret, shuts down its http endpoint, populates the secret with the cluster-admin kubeconfig and kubeadmin password, and exits.
- The user reads the secret.

Such an exchange can be easily automated. This may or may not be an easier approach, but it seems worth at least considering and listing as an alternative.